Fix some bugs in the regression tests and add more regression tests

JohnoKing opened the pull request:
src/cmd/ksh93/tests/{basic.sh,builtins.sh,shtests}:
- Redirect error output from the `ulimit` builtin to silence irrelevant errors in the regression tests (these errors may occur when a command such as `ulimit -t 4` is run before the regression tests).
- Shellquote the error messages from the `getconf` regression tests.

src/cmd/ksh93/tests/{arrays,io,variables}.sh:
- Backport the ksh2020 regression tests for the following bugs:
  <https://github.com/att/ast/issues/23>
  <https://github.com/att/ast/issues/203>
  <https://github.com/att/ast/issues/472>
  <https://github.com/att/ast/issues/492>

src/cmd/ksh93/tests/{basic,pty}.sh:
- Add regression tests for the following bugs:
  <https://github.com/att/ast/issues/1160>
  <https://github.com/att/ast/issues/1461>
- The buggy [ksh2020 fix for `[ -t 1 ]`](https://github.com/att/ast/pull/1083) in non-forking command substitutions caused `[ -t 1 ]` to break in non-comsub virtual subshells (when used in interactive shells). Reproducer (which this pull request adds to the regression tests):
```sh
$ ( [ -t 1 ]; echo $? )
1  # Always fails
```

JohnoKing committed:
Fix some bugs in the regression tests and add more regression tests

src/cmd/ksh93/tests/{basic.sh,builtins.sh,shtests}:
- Redirect error output from the ulimit builtin to silence irrelevant
  errors in the regression tests (these errors may occur when a
  command such as 'ulimit -t 4' is run before the regression tests).
- Shellquote the error messages from the getconf regression tests.

src/cmd/ksh93/tests/{arrays,io,variables}.sh:
- Backport the ksh2020 regression tests for the following bugs:
  <https://github.com/att/ast/issues/23>
  <https://github.com/att/ast/issues/203>
  <https://github.com/att/ast/issues/472>
  <https://github.com/att/ast/issues/492>

src/cmd/ksh93/tests/{basic,pty}.sh:
- Add regression tests for the following bugs:
  <https://github.com/att/ast/issues/1160>
  <https://github.com/att/ast/issues/1461>
- The ksh2020 fix for [ -t 1 ] in non-forking command substitutions
  caused the following bug in interactive shells:
    $ ( [ -t 1 ]; echo $? )
    1  # Always fails
  To avoid introducing this bug, this commit adds a regression
  test for it.

JohnoKing committed:
Minor tweak to error message

JohnoKing committed:
Merge remote-tracking branch 'upstream/master' into extra-regress-tests

McDutchie commented on src/cmd/ksh93/tests/basic.sh:
```diff
@@ -832,5 +832,21 @@ got="$(join <(printf '%d\n' 1 2) <(printf '%d\n' 1 2))"
+# function body to stdout.
+# https://github.com/att/ast/issues/1160
+got="$($SHELL -c '
+	for i in $(seq 1024)
```
A few things on this one:
* The `seq` command is non-standard and unportable. Please use an arithmetic `for` loop instead.
* I think this test would be better placed in functions.sh.
* I can't get this test to fail on my macOS system when running it on 93v- or ksh2020 (using `bin/shtests -p basic SHELL=/usr/local/bin/ksh2020`, etc.).

JohnoKing committed:
Fix issues with eval function body test

- Move the test to the start of functions.sh (if it's at the end
  of the script, it refuses to fail under ksh2020). Output from
  this regression test when run against ksh2020:
    $ bin/shtests -p functions
    #### Regression-testing /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64/bin/ksh ####
    test functions begins at 2021-05-02+21:22:19
            functions.sh[46]: eval'ing function dumps function body to stdout (got $' { eval "bar() { FAILURE; }"; }\n { FAILURE; }')
- Use arithmetic for loop instead of seq.

JohnoKing committed:
Minor fix to POSIX mode regression tests in ksh93v-

In ksh93v- [[ -o ?posix ]] doesn't return an error (because it's
implemented in the bash mode). However, 'set -o posix' will fail in
ksh93v- if it's not in bash compatibility mode, which causes this
test script to exit prematurely.
